autokeras/tuner.py

```diff
@@ -72,27 +68,71 @@ def load_trial(self, trial):
         path = os.path.join(trial.directory, filename)
         self.hypermodel.load_preprocessors(path)
 
-    @staticmethod
-    def add_earlystopping_callback(callbacks):
-        if not callbacks:
-            callbacks = []
-
-        try:
-            callbacks = copy.deepcopy(callbacks)
-        except:
-            raise ValueError(
-                'All callbacks used during a search '
-                'should be deep-copyable (since they are '
-                'reused across executions). '
-                'It is not possible to do `copy.deepcopy(%s)`' %
-                (callbacks,))
-
-        if not [callback for callback in callbacks
-                if isinstance(callback, tf.keras.callbacks.EarlyStopping)]:
-            # The patience is set to 30 based on human experience.
-            callbacks.append(tf.keras.callbacks.EarlyStopping(patience=30))
-
-        return callbacks
+    def search(self, *fit_args, **fit_kwargs):
+        # Format the arguments.
+        fit_kwargs.update(
+            dict(zip(inspect.getfullargspec(tf.keras.Model.fit).args, fit_args)))
+        fit_args = []
+        fit_kwargs.setdefault('callbacks', [])
+        if fit_kwargs['callbacks'] is None:
+            fit_kwargs['callbacks'] = []
+        fit_kwargs.setdefault('epochs', None)
+
+        # Insert early-stopping permanently when epochs is None.
+        callbacks = fit_kwargs['callbacks']
+        if fit_kwargs['epochs'] is None:
+            fit_kwargs['epochs'] = 1000
+            if not any([isinstance(callback, tf.keras.callbacks.EarlyStopping)
+                        for callback in callbacks]):
+                callbacks = callbacks + [
+                    tf.keras.callbacks.EarlyStopping(patience=30)]
+        fit_kwargs['callbacks'] = callbacks
+
+        # Insert early-stopping temporarily for the search.
```

Comment:
I don't understand the logic here. `final_fit` is going to be True IFF there is no EarlyStopping callback. But the only case where there is no EarlyStopping callback is if `fit_kwargs['epochs']` is set by the user *and* the user did not pass an EarlyStopping callback.

Please clarify what the intent is, and why.

Reply:
Yes, this is what we want. If the user specified the epochs and did not put EarlyStopping in the callbacks, it means the user wants the model to be trained to the specific epochs. In this case, we need the final fit to fully train the model, and will not use early stopping in the final fit.

This temporary early stopping is just for accelerating the search. It is not related to the case which user did not specify the epochs.

For the case of user not specifying epochs or user specified epochs and put earlystopping in  the callbacks, each trial would be run exactly according to the user specifications. So we don't need the final fit.

Reply:
I would suggest refactoring this part of the code to make the logic easier to follow (e.g. using an enumeration of cases), and adding comments.